ledger/accountdb.go

```diff
@@ -1026,74 +1080,395 @@ func accountsPutTotals(tx *sql.Tx, totals ledgercore.AccountTotals, catchpointSt
 	return err
 }
 
-// accountsNewRound updates the accountbase and assetcreators tables by applying the provided deltas to the accounts / creatables.
-// The function returns a dbAccountData for the modified accounts which can be stored in the base cache.
-func accountsNewRound(tx *sql.Tx, updates compactAccountDeltas, creatables map[basics.CreatableIndex]ledgercore.ModifiedCreatable, proto config.ConsensusParams, lastUpdateRound basics.Round) (updatedAccounts []dbAccountData, err error) {
+func loadHoldings(stmt *sql.Stmt, eah ledgercore.ExtendedAssetHolding) (map[basics.AssetIndex]basics.AssetHolding, ledgercore.ExtendedAssetHolding, error) {
+	if len(eah.Groups) == 0 {
+		return nil, ledgercore.ExtendedAssetHolding{}, nil
+	}
+	var err error
+	holdings := make(map[basics.AssetIndex]basics.AssetHolding, eah.Count)
+	for gi, g := range eah.Groups {
+		holdings, eah.Groups[gi], err = loadHoldingGroup(stmt, g, holdings)
+		if err != nil {
+			return nil, ledgercore.ExtendedAssetHolding{}, err
+		}
+	}
+	eah.SetLoaded()
+	return holdings, eah, nil
+}
 
-	var insertCreatableIdxStmt, deleteCreatableIdxStmt, deleteByRowIDStmt, insertStmt, updateStmt *sql.Stmt
+func loadHoldingGroup(stmt *sql.Stmt, g ledgercore.AssetsHoldingGroup, holdings map[basics.AssetIndex]basics.AssetHolding) (map[basics.AssetIndex]basics.AssetHolding, ledgercore.AssetsHoldingGroup, error) {
+	if holdings == nil {
+		holdings = make(map[basics.AssetIndex]basics.AssetHolding, g.Count)
+	}
+	groupData, err := loadAssetHoldingGroupData(stmt, g.AssetGroupKey)
+	if err != nil {
+		return nil, ledgercore.AssetsHoldingGroup{}, err
+	}
+	aidx := g.MinAssetIndex
+	for i := 0; i < len(groupData.AssetOffsets); i++ {
+		aidx += groupData.AssetOffsets[i]
+		holdings[aidx] = groupData.GetHolding(i)
+	}
+	g.Load(groupData)
+	return holdings, g, nil
+}
 
-	deleteByRowIDStmt, err = tx.Prepare("DELETE FROM accountbase WHERE rowid=?")
+func loadAssetHoldingGroupData(stmt *sql.Stmt, key int64) (group ledgercore.AssetsHoldingGroupData, err error) {
+	var buf []byte
+	err = stmt.QueryRow(key).Scan(&buf)
 	if err != nil {
+		return ledgercore.AssetsHoldingGroupData{}, err
+	}
+
+	err = protocol.Decode(buf, &group)
+	return
+}
+
+func accountsNewCreate(qabi *sql.Stmt, qaei *sql.Stmt, addr basics.Address, ad basics.AccountData, proto config.ConsensusParams, updatedAccounts []dbAccountData, updateIdx int) ([]dbAccountData, error) {
+	assetsThreshold := config.Consensus[protocol.ConsensusV18].MaxAssetsPerAccount
```

Comment:
This need to be based on the current round protocol version. This makes things harder, since it means that we'll need to adjust the committedUpTo() so that it would break updates on protocol boundaries.